### ISSUE TITLE
[Templates v2] Add a template variable that replaces `dateFormat`

### ISSUE THREAD

**Issue by movermeyer** [foambubble/foam#564]
As described in #534, `Foam: Open Daily Note` uses the `foam.openDailyNote.filenameFormat` setting and passes it to `dateFormat`. This allows users the flexibility to put the current timestamp into the daily notes filename using any format available via `dateFormat`. This is much more flexible than what is available in VSCode Snippet variables.

Add a new syntax / mechanism to maintain the flexibility of `dateFormat` within a template variable.

**Comment by movermeyer** [comment]
Instead of doing anything complicated in the snippet variable syntax (which I want to improve so we exactly match the VSCode/TextMate syntax), what if we introduce a new settings variable (perhaps `foam.dateFormatVariablePattern`) that a `${FOAM_DATE_FORMAT}` snippet variable uses to pass into `dateFormat` to get its value?

```json
"foam.dateFormatVariablePattern": "yyyy-mm-dd"
```

**Comment by riccardoferretti** [comment]
mmm.. let's see if I understand:
- add a new setting: `foam.dateFormatVariablePattern`. This uses the `dateFormat` mechanism to populate the variable
- make the setting available as a variable in the foam template (kinda like `${FOAM_TITLE}`

IIUC I wonder if this solution would be a bit complicated to handle for the user.

Can you walk me through some use cases you are solving for?

**Comment by movermeyer** [comment]
>    add a new setting: foam.dateFormatVariablePattern. This uses the dateFormat mechanism to populate the variable
>    make the setting available as a variable in the foam template (kinda like ${FOAM_TITLE}

Yeah, that's correct.

---

> I wonder if this solution would be a bit complicated to handle for the user. Can you walk me through some use cases you are solving for?

Users can already use the [`CURRENT_YEAR`/`CURRENT_MONTH`, etc.](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables) offered natively by VS Code to insert date/time values in templates.

IMO, this is probably enough flexibility to represent date times in notes:

* `2021-05-01` (`$CURRENT_YEAR-$CURRENT_MONTH-$CURRENT_DATE`)
* `05/01/2021` (`$CURRENT_MONTH/$CURRENT_DATE/$CURRENT_YEAR`)
* `01.05.2021` (`$CURRENT_DATE.$CURRENT_MONTH.$CURRENT_YEAR`)
* `Saturday May 1, 2021` (`$CURRENT_DAY_NAME $CURRENT_MONTH_NAME $CURRENT_DATE, $CURRENT_YEAR`)

However, users that are currently using `foam.openDailyNote.filenameFormat` have access to any options that [`dateFormat` supports](https://github.com/felixge/node-dateformat#mask-options).

* Example: `Week 17: 2021-05-01 GMT-0500` (`'Week' WW: yyyy-mm-dd Z`)

However, in #534, we discussed maintaining this flexibility. So we need a new syntax that they can migrate to.

### Option 1: "Method call" syntax

My first thought was to extend the variable syntax to support a syntax that "looks like a method call":

```markdown
# ${FOAM_DATE_FORMAT("'my-note'-yyyy-mm-dd")}
```

(Aside: We'd use double-quotes, in order to avoid conflicting with dateFormat's use of single-quotes. If we needed to support single quotes as well, we'll need to support escaping (likely using `\`))

**Benefits**

* It's all in-line; there's no "magic" hidden from the user

**Detriments**

* The variable no longer follows [VSCode's Snippet grammar](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_grammar)

When implementing the [first version of Foam-specific variable](https://github.com/foambubble/foam/pull/549) (`${FOAM_TITLE}`), for simplicity's sake I didn't implement support for:

* the defaulting feature (eg. `${variable:default}`)
* the format/transform feature (eg. `${variable/(.*)/${1:/upcase}/}`).

which are available to other VScode snippet variables. I was holding out hope that in the future, we could adopt (or vendor) an off-the-shelf snippet parser to enable these features in Foam-specific variables. It would handle both the parsing of the variables, and the applying of transforms (Unfortunately, the one built into VSCode is not exposed for use).

If we were to extend the variable syntax, then we would no longer be able to use a stock snippet parser, and instead have to implement our own custom one.

### Option 2: "Settings lookup" syntax

This is the option I proposed in my [previous comment](https://github.com/foambubble/foam/issues/564#issuecomment-826318379).

```markdown
# my-note-${FOAM_DATE_FORMAT}
```

```json
"foam.dateFormatVariablePattern": "yyyy-mm-dd"
```

**Benefits**

* It still fits within the VSCode snippet variable grammar

**Detriments**

* The value used is hidden in the `foam.dateFormatVariablePattern`
* All instances of `${FOAM_DATE_FORMAT}` get the same value, you cannot use multiple different custom date formats.

### Option 3: TDB

Are there other good options available?

## Summary

Any of these choices are much more compact than using the built-in variables, so we might expect that users might abandon the built-in variables for this new syntax.

If we decide we are going to write our own VSCode snippet parser anyway, the "method call" syntax might be preferable.

### How likely are we to implement our own snippet parser?

The snippet variable syntax is not designed to be done in multiple passes. Attempting to do so leads to bugs.
In order to fix #602, we would need our own snippet parser.

At that point, it "probably" isn't that much more difficult to extend the syntax to support the "method call" syntax.

---

(ie. after writing this comment, and finding #602, I think I prefer the "method call" syntax)

**Comment by riccardoferretti** [comment]
thanks for the thorough explanation, I agree with everything you say.
I am thinking of two alternative approaches, curious to know your reaction to them.

1. We could combine the method call syntax that VSCode exposes with the dateFormat feature.

`${variable/(.*)/${1:/dateFormat(dd-mm}/})`

This might have a couple of benefits:
- we could look at the VS Code source responsible for this parsing and embed it into Foam (if it can't be extended, which I don't think is the case)
- the syntax would be compatible with VS Code variable substitution, and would simply be ignored e.g. when the values provided are incorrect (kinda like what VS Code would do if you provide an invalid method call)
- keeps everything in one place, while keeping the flexibility as well

That being said, I do wonder (and unfortunately we don't have analytics to know that) how much `dateFormat` specific features are used by users, we are risking building something that has some complexity in it and very few people, if anyone, uses.

2. We could *not* support `dateFormat` in our template path, but keep for backwards compatibility the current settings that support `dateFormat`. If the `daily-note.md` template provides a path, use that. if not, fallback to the old settings. if they don't exist, use defaults.

thoughts?
